### ISSUE TITLE
Questions about coordinates axis and pose bounds

### ISSUE THREAD

**Issue by LZL-CS** [ActiveVisionLab/direct-posenet#4]
Hi, I have some questions about your released code:

1. In your code, you correct the rotation matrix from **up left forward** to **up right backward**, may I know why the 7Scenes dataset coordinate axis looks like this? and why do you need to correct the axis? 
https://github.com/ActiveVisionLab/direct-posenet/blob/1dfbe32125878402eda551950867086b1e3010ab/dataset_loaders/load_7Scenes.py#L323
I understand that for the NeRF (LLFF dataset), we need to correct the rotation matrix from OpenGL **right up backward** to COLMAP OpenCV **right down forward**. 
![coordinate](https://user-images.githubusercontent.com/54640259/190970599-1230cbb0-793a-43db-94ab-ed356449e025.jpeg)

2. Why need to correct the axis using `M*([R|T]*M)`? how to derivate it?
https://github.com/ActiveVisionLab/direct-posenet/blob/1dfbe32125878402eda551950867086b1e3010ab/dataset_loaders/load_7Scenes.py#L331

3. You set the pose bounds for `near=0.5` and `far =2.5`, may I know why you choose these values? Is this value the same for 7 scenes (heads, stairs, chess, ...)?

Hope for your reply, thank you very much!

**Comment by chenusc11** [comment]
Q. 1. & 2.:
Thanks for your question. Since I didn't know which coordinate system 7-Scenes dataset was using, I tried to re-align the 7-Scenes GT poses to the NeRF's myself. The way I did it is to align 7-scenes dataset's coordinate -> colmap coordinates -> NeRF coordinates.

It looks like something like this (Note the camera frustum should also be aligned even though I didn't show it in the picture):
<img width="427" alt="Screen Shot 2022-09-22 at 3 30 27 PM" src="https://user-images.githubusercontent.com/8814381/191775364-32a47ecf-22b3-41b1-9c51-1f60fc8d66c6.png">

all_poses = flip_M@(all_poses@flip_M) was just a heuristic combination that works correctly for camera position and camera frustum's orientation to training a NeRF model.

Fortunately, we only need to do it once despite being a tedious process to align dataset.

Q.3.:
The near/far bounds were manually selected by estimating the bounds to the camera. As we stated  in the paper, we set 7-scene's near and far bounds [bn, bf] to [0.5, 4] except for the Heads, which was [0.5, 2.5]

I recall there's a paper summarizing the scene's size for 7-Scenes though I don't remember exactly which paper.

<img width="293" alt="Screen Shot 2022-09-22 at 3 47 22 PM" src="https://user-images.githubusercontent.com/8814381/191779081-dafcf1f0-df00-4939-98e8-991db3149b79.png">

**Comment by LZL-CS** [comment]
Hi @chenusc11, thanks a lot for your detailed reply, but I am still confused about some issues:

To Q1:
From your answer, we know that you align the **7-scenes dataset's coordinates (unkonw)** -> **colmap coordinates (right, down, forward)** -> **NeRF coordinates (right, up, backward)**, which follows operations as below:
(1) `all_poses = flip_M @ (all_poses @ flip_M)  # bug here M*([R|T]*M)` 
(2) `dirs = torch.stack([(i - W * .5) / focal, -(j - H * .5) / focal, -torch.ones_like(i)], -1)`

While in the NeRF, the author align the **llff dataset's coordinates (down, right, backward)** -> **OpenGL (right, up, backward)** -> **COLMAP (right, down, forward)**, which follows operations as below  (refer to [nerf_pl](https://github.com/kwea123/nerf_pl/)) :
(1) `poses = np.concatenate([poses[..., 1:2], -poses[..., :1], poses[..., 2:4]], -1)`
(2) `directions = torch.stack([(i-cx)/fx, -(j-cy)/fy, -torch.ones_like(i)], -1)`

Does it mean that your final coordinates are different from NeRF used? 

To Q2:
From my naive idea, the `flip_M` is the rotation matrix used to transfer coordinates **x, y, z** to **x, -y, -z**, thus `all_poses @ flip_M` can simply do it, but why do you use `flip_M` left multiply `all_poses @ flip_M` again?

**Comment by chenusc11** [comment]
Q1:
I'm not familiar with nerf_pl's implementation since my NeRF was based on the [nerf-pytorch](https://github.com/yenchenlin/nerf-pytorch) repo. This repo was an exact pytorch replicate from the [official NeRF](https://github.com/bmild/nerf), which was in Tensorflow.

I believe my 7-scene's dataset is converted to exactly the same as what the official NeRF/nerf-pytorch is using. 

In our actual fix_coord() implementation, it is a direct transform from 7-scenes dataset's coordinates (unknown) -> NeRF coordinates, in which we omit the intermediate step (colmap coord.). I think I could fix the comments to make it more clear.

Q2:
Although it's been a long time since I was implementing this, I think this was inspired by [mirror matrices](https://wp.optics.arizona.edu/optomech/wp-content/uploads/sites/53/2016/08/6-Mirror-matrices.pdf). When I was aligning my coordinate system with NeRF's, I found out that I need to take a mirror-like transformation in order to make the camera ray shooting at the correct direction.

In this case, flip_M on the right side is actually flip_M.T. But since the flip_M and flip_M.T are the same, I ignore the transpose operation.

I think it will make more sense if you can draw the camera rays to see what happens before and after the conversion. For example: draw something like this, and see what happens before and after transformation.
![Screen Shot 2022-09-23 at 1 21 29 PM](https://user-images.githubusercontent.com/8814381/191959017-d38a80cc-8fb9-4877-b449-6919f9e28fa7.png)

**Comment by LZL-CS** [comment]
Very thanks for your patience and reply again,  will chek it soon.
